Implement XGBoost for classification and regression

michaellevy opened the pull request:
Added xgboost, removed knn. Played with hyperparameter values/ranges a bit for default and random-search, but only on pima_diabetes and small synthetic datasets, so one thing that would be useful would be to train on a larger real dataset and see if the hyperparameter values make sense there. Otherwise, I don't think there's a ton to review, no new functions or anything. Will update vignettes in #1148.

mmastand commented on R/models.R:
```diff
@@ -11,15 +11,19 @@
+#'   \item{max_depth: Maximum tree depth, [0, Inf]. Larger means more complex
+#'   models and so greater likelihood of overfitting. 0 produces no limit on depth.}
+#'   \item{subsample: Fraction of data to use in each training instance, (0, 1].}
+#'   \item{colsample_bytree: Fraction of features to use in each tree, (0, 1].)}
```
extra paren

mmastand commented on R/models.R:
```diff
@@ -11,15 +11,19 @@
+#'   \item{min_child_weight: Minimum sum of instance weight need to keep partitioning,
+#'   [0, Inf]. Larger values mean more conservative models.}
+#'   \item{nrounds: Number of rounds of boosting, [0, Inf)}
 #' }
```
It would be great to have a comment about relative model training time in the `machine_learn` docs to give people a ballpark. Something like, GLM is probably fastest, RF is next, xgb is slowest. This could also live in the `models` file.

michaellevy replied:
Good idea. I'll do that in #1107

mmastand commented on R/models.R:
```diff
@@ -11,15 +11,19 @@
+#'   \item{colsample_bytree: Fraction of features to use in each tree, (0, 1].)}
+#'   \item{min_child_weight: Minimum sum of instance weight need to keep partitioning,
+#'   [0, Inf]. Larger values mean more conservative models.}
+#'   \item{nrounds: Number of rounds of boosting, [0, Inf)}
```
Please put a comment here saying that higher `nrounds` leads to down the path of the overfittingside.

mmastand commented on tests/testthat-1.R:
```diff
@@ -1,4 +1,4 @@
 library(healthcareai)
 Sys.setenv("R_TESTS" = "")
-test_check("healthcareai", filter = "^[(a-m)|(A-M)]")
+test_check("healthcareai", filter = "^[(a-i)|(A-I)]")
```
Why do we now have 3 testthat files?

michaellevy replied:
@mmastand This is to avoid the Travis time out errors. If travis doesn't get anything back from R in ten minutes, it fails the build. When one of these test files completes, it sends an update to Travis.

michaellevy commented:
@mmastand thanks -- good review and suggestions. I added support for XGB variable importance; it sounded like you wanted to take another look at that.

I added a note to models and tune_models docs about xgb benefiting from greater tune_depth. I'm going to hold off on all the model training time until the speed tests finish tomorrow. 

The x-axis labels on `plot.model_list` is kind of intentional. At least it's that way instead of on every individual plot intentionally, but this is the first time we use multiple columns within a model, so it does look missing now. There's still that nagging vertical spacing issue. I'd like to leave it alone for now and switch to [patchwork](https://github.com/thomasp85/patchwork) from cowplot once it's on CRAN.

mmastand commented:
Variable importance looks good, notes look good, agreed that it's the 2-column grid that's causing issues. I'm happy with leaving it.

Approved!
